builder: Automatically compress man pages. Resolves #16

joebonrichie opened the pull request:
Compress any man page files found in the manifest with zstd.

There are three considerations:-
- It _will_ increase the size of the resulting .stone file(s), compressing .stones with --long could resolve this.
- Info pages are currently ignored but could also be compressed.
- Updating symlinks is not yet tested.

joebonrichie commented:
Just need to find a file with man page symlinks now and should be good to go. There is also an annoying bug where `formattedSize` introduces some useless whitespace.

sunnyflunk commented:
> Just need to find a file with man page symlinks

Probably easier to add a symlink manually to test. But I'm pretty sure it won't work. The first check ensures only Regular files hit the NextFunction (i.e. so a Symlink will never hit compressPage).

Also with `if ((fileInfo.type == FileType.Symlink) && (fileInfo.type != FileType.Directory))` the 2nd check is redundant as if it's a Symlink, then it will always not be a Directory.

joebonrichie commented:
This is borked with the new zstd pledged size stuff. With the initial move to zstdoubledee it was working okay. Just hangs forever when emitting the packages. Don't really know what's going on.

joebonrichie commented:
Bisected to here: https://github.com/serpent-os/moss-format/commit/71bbd9735d40e277ab728a5c44fa1c2c36541e6b

joebonrichie commented:
The st_size of the compressed man pages is not known and defaults to 0
```
[11:09:56] TRACE     FILE: digest [187, 208, 173, 55, 192, 114, 7, 140, 2, 179, 169, 103, 64, 132, 226, 134], path /var/cache/boulder/build/nano-7.1-4/install/usr/share/man/man1/rnano.1.zst, st_size 0
[11:09:56] TRACE     zstd: pledged size 2.47 MiB
```

It seems to hang during `contentPayload.addFile(file.digest, file.fullPath, file.stat.st_size);` when it encounters the first compressed man page.

joebonrichie commented:
Seems to be failing to flush (remaining never decreases)
`[12:57:12] TRACE     FLUSHING remaining 18446744073709551544`

from here: https://github.com/serpent-os/moss-format/blob/main/source/moss/format/binary/writer/zstd_token.d#L117

joebonrichie commented:
Switched to gzip compression and it's now working, meaning `mandb` can correctly display the resulting files! Switching to zstd in the future when the high level API is a bit more fleshed out is an option but I don't really want to use the low-level bindings for something like this. 

Updated symlinks is now seemingly working (with the `xz` pkg) but I need to find a package I can safely remove and reinstall in the rootfs to fully test it, which is the last blocker.
`[Man] Updated symlink /usr/share/man/ro/man1/unxz.1 to xz.1.gz`
